File: src/main/java/me/alexirving/bot/commands/Link.java
package me.alexirving.bot.commands;

import com.mattmalec.pterodactyl4j.PteroBuilder;
import com.mattmalec.pterodactyl4j.client.entities.PteroClient;
import me.alexirving.bot.Database;
import me.alexirving.bot.ServerManager;
import me.alexirving.bot.utils.Utils;
import me.alexirving.bot.utils.Messages;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;

import java.io.IOException;
import java.util.Objects;

import static me.alexirving.bot.Database.memberPteroClients;
import static me.alexirving.bot.Database.memberTokens;
import static me.alexirving.bot.utils.Utils.updateActivity;

public class Link {
    public void linkCommand(SlashCommandEvent event){
        Member member = event.getMember();
        assert member != null;
        String memberId = member.getId();
        String guildId = Objects.requireNonNull(event.getGuild()).getId();
        if (Utils.isLinkValid(Utils.getConfigMessage(guildId, Messages.API_URL))){
        PteroClient pteroClientLink = PteroBuilder.createClient(Utils.getConfigMessage(guildId, Messages.API_URL), Objects.requireNonNull(event.getOption("key")).getAsString());
        memberPteroClients.put(guildId, memberId, pteroClientLink);
        pteroClientLink.retrieveAccount().executeAsync(ac -> {
            memberTokens.put(guildId, memberId, Objects.requireNonNull(event.getOption("key")).getAsString());
            try {
                Database.updateTokensJson();
            } catch (IOException e) {
                e.printStackTrace();
            }
            updateActivity(event.getJDA());
            event.getHook().editOriginal(Utils.getConfigMessage(guildId, Messages.LINKED)).queue();


        }, throwable -> event.getHook().editOriginal(Utils.getConfigMessage(guildId, Messages.INVALID_APIKEY)).queue());
    }else{
            event.getHook().editOriginal("API URL NOT SETUP CORRECTLY!").queue();
        }
    }
}

File: src/main/kotlin/me/alexirving/bot/command/commands/Link.kt
package me.alexirving.bot.command.commands

import com.mattmalec.pterodactyl4j.PteroBuilder
import me.alexirving.bot.PteroBot
import me.alexirving.bot.command.Argument
import me.alexirving.bot.command.Command
import me.alexirving.bot.utils.addRow
import me.alexirving.bot.utils.setValue
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent
import net.dv8tion.jda.api.interactions.commands.OptionType

class Link(val bot: PteroBot) : Command(
    "link",
    "Link your account to pterodactyl panel account!",
    Permission.MESSAGE_SEND,
    Argument(OptionType.STRING, "key", "Your api key!", true)
) {
    override fun execute(event: SlashCommandEvent) {
        event.deferReply().setEphemeral(true).queue()
        val userId = event.user.id

        if (bot.configs[event.guild]!!.getValue("url") == "default") {
            event.hook.editOriginal("An error occurred, this server was not setup correctly!").queue()
            return
        }

        //Adding user to the database if they aren't already in it.
        addRow("${bot.clientId}_keys", userId)

        //Not using session manager since we don't want it to set if it failed.
        try {
            val tempClient =
                PteroBuilder.create(bot.configs[event.guild]!!.getValue("url"), event.getOption("key")!!.asString)
                    .setHttpClient(bot.okHttpClient).buildClient()
            tempClient.retrieveAccount()
                .executeAsync({
                    event.hook.editOriginal("Linked well!").queue()
                    setValue("${bot.clientId}_keys", userId, event.guild!!.id, event.getOption("key")!!.asString)
                    bot.sessionManager.setClient(event.member!!, tempClient)
                }, {
                    event.hook.editOriginal("API key was invalid!").queue()
                })
        } catch (ignored: Exception) {
            event.hook.editOriginal("An error occurred, this server was not setup correctly!").queue()
        }


    }

}